notebooks/general_tips_2022-11.ipynb

```diff
@@ -0,0 +1,879 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def hypotenuse(x, y):\n",
+    "    return (x*x + y*y) ** 0.5"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "137 ns ± 4.45 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)\n"
+     ]
+    }
+   ],
+   "source": [
+    "%timeit hypotenuse(3.0, 4.0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### External Packages"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import math"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "59.1 ns ± 2.15 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)\n"
+     ]
+    }
+   ],
+   "source": [
+    "%timeit math.hypot(3.0, 4.0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Additional Tip: Avoid the dot operation!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from math import hypot"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "40.3 ns ± 1.26 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)\n"
+     ]
+    }
+   ],
+   "source": [
+    "%timeit hypot(3.0, 4.0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data types/structures"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Built-in Types: List versus Dictionary"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "\n",
+    "needles = [2, 3, 10, 99, 101]\n",
+    "haystack_list = list(np.random.randint(low=0, high=100, size=(1_000_000)))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "19.1 ms ± 420 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%timeit\n",
+    "needles_found = 0\n",
+    "for needle in needles:\n",
+    "    if needle in haystack_list:\n",
+    "        needles_found += 1"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "haystack_dict = {value: key for key, value in enumerate(haystack_list)}"
```

Comment:
Something rather sneaky happens here. 

Because we're creating a dictionary that has to have unique keys by having it use `value` as the key for each item it actually creates a far smaller dictionary than the original list. This is because there are only 100 unique `value` instances so it discards all other items when enumerating through your 1_000_000 item list. 

I think it might be worth switching this example to use string matching instead, like so:

```python
str_list = ["foo","bar","magic","MAGIC","extra","new","fun","place","lemon", "leMon","ferrai"]

wanted_list = ["foo","bar","new"]
```
```python
%%timeit 
wanted_no = 0
for item in wanted_list:
    if item in str_list:
        wanted_no += 1
```
Then make a dictionary:
```python
wanted_dict = {key: value for key, value in enumerate(str_list)}
```
```python
%%timeit 
wanted_no = 0
for item in wanted_list:
    if item in wanted_dict:
        wanted_no += 1
```

We have to use the `if item in wanted_dict` line because that uses hashing for comparison (and should be performant) whilst `if item in wanted_dict.keys()` is actually the same as comparison to a list.

What do you think?

Reply:
I agree with you, but your solution is not working as well.
If you test, the `wanted_no` is going to be different in both examples.
I will try fix this in a way that boot the list/dict are equivalent as well the result.

Reply:
You're quite right! It's because i've got my keys and values turned around.

If you do:

```python
wanted_dict = {value: key for key, value in enumerate(str_list)}
```

That should ensure the strings are the keys so the `in` test works.

Reply:
Although for me this isn't actually faster than the list. This may be because the list is quite small

Reply:
Lets keep it simple.

I will change this:

`haystack_list = list(np.random.randint(low=0, high=100, size=(1_000_000)))`

by:

`haystack_list = list(range(2,10,2))`

By doing this, we can easily change the size and ensure that it will not have any duplicate values, So we can keep my first version of the problem.

We also can play with length and show that for small lists, the result is very close, but as we increase the size the performance changes.